// nocodb-sdk/build/module/lib/columnHelper/columns/Number.js
import { SilentTypeConversionError } from '../../../lib/error';
import { parseIntValue, serializeIntValue } from '..';
import AbstractColumnHelper from '../column.interface';
import { toSafeInteger } from '../../../lib/helperFunctions';
import { populateFillHandleStringNumber } from '../utils/fill-handler';
export class NumberHelper extends AbstractColumnHelper {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            isLocaleString: false,
        };
    }
    serializeValue(value, params) {
        value = serializeIntValue(value);
        if (value === null) {
            if (params.isMultipleCellPaste) {
                return null;
            }
            else {
                throw new SilentTypeConversionError();
            }
        }
        return toSafeInteger(value);
    }
    parseValue(value, params) {
        // Return empty string for null/undefined values to prevent "null" text when pasting
        if (value === null || value === undefined) {
            return '';
        }
        return parseIntValue(value, params.col);
    }
    parsePlainCellValue(value, params) {
        var _a;
        return `${(_a = parseIntValue(value, params.col)) !== null && _a !== void 0 ? _a : ''}`;
    }
    // using string number fill handler
    populateFillHandle(params) {
        return populateFillHandleStringNumber(params);
    }
}
