// nocodb-sdk/build/main/lib/columnHelper/columns/Duration.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DurationHelper = void 0;
const error_1 = require("../../../lib/error");
const __1 = require("..");
const column_interface_1 = __importDefault(require("../column.interface"));
class DurationHelper extends column_interface_1.default {
    constructor() {
        super(...arguments);
        this.columnDefaultMeta = {
            duration: 0,
        };
    }
    serializeValue(value, params) {
        value = (0, __1.serializeDurationValue)(value, params.col);
        if (value === null) {
            if (params.isMultipleCellPaste) {
                return null;
            }
            else {
                throw new error_1.SilentTypeConversionError();
            }
        }
        return value;
    }
    parseValue(value, params) {
        return (0, __1.parseDurationValue)(value, params.col);
    }
    parsePlainCellValue(value, params) {
        var _a;
        return (_a = (0, __1.parseDurationValue)(value, params.col)) !== null && _a !== void 0 ? _a : '';
    }
}
exports.DurationHelper = DurationHelper;
