// nocodb-sdk/build/module/lib/columnHelper/utils/parser.js
import dayjs from 'dayjs';
import { convertMS2Duration } from '../../../lib/durationUtils';
import { parseProp, roundUpToPrecision } from '../../../lib/helperFunctions';
import { ncIsArray, ncIsBoolean, ncIsNaN, ncIsNull, ncIsNumber, ncIsObject, ncIsString, ncIsUndefined, } from '../../../lib/is';
import { constructTimeFormat } from '../../../lib/dateTimeHelper';
import { checkboxTypeMap } from '../../../lib/columnHelper/utils/common';
export const parseDefault = (value) => {
    var _a;
    try {
        if (typeof value === 'object') {
            return JSON.stringify(value);
        }
        else {
            return (_a = value === null || value === void 0 ? void 0 : value.toString()) !== null && _a !== void 0 ? _a : null;
        }
    }
    catch (_b) {
        return null;
    }
};
export const parseIntValue = (value, col) => {
    if (ncIsNaN(value)) {
        return null;
    }
    value = parseInt(value.toString(), 10);
    const columnMeta = parseProp(col === null || col === void 0 ? void 0 : col.meta);
    if (columnMeta.isLocaleString) {
        return Number(value).toLocaleString();
    }
    return Number(value);
};
export const parseDecimalValue = (value, col) => {
    var _a, _b, _c, _d;
    if (ncIsNaN(value)) {
        return null;
    }
    const columnMeta = parseProp(col.meta);
    if (columnMeta.isLocaleString) {
        return Number(roundUpToPrecision(Number(value), (_a = columnMeta.precision) !== null && _a !== void 0 ? _a : 1)).toLocaleString(undefined, {
            minimumFractionDigits: (_b = columnMeta.precision) !== null && _b !== void 0 ? _b : 1,
            maximumFractionDigits: (_c = columnMeta.precision) !== null && _c !== void 0 ? _c : 1,
        });
    }
    return roundUpToPrecision(Number(value), (_d = columnMeta.precision) !== null && _d !== void 0 ? _d : 1);
};
export const parsePercentValue = (value) => {
    if (ncIsNaN(value)) {
        return null;
    }
    return `${Number(value)}%`;
};
export const parseDurationValue = (value, col) => {
    var _a;
    const durationType = ((_a = parseProp(col.meta)) === null || _a === void 0 ? void 0 : _a.duration) || 0;
    return convertMS2Duration(value, durationType);
};
export const parseCheckboxValue = (value) => {
    if (ncIsBoolean(value))
        return value;
    if (ncIsString(value)) {
        const strval = value.trim().toLowerCase();
        const parsedValue = checkboxTypeMap[strval];
        if (!ncIsNull(parsedValue) && !ncIsUndefined(parsedValue)) {
            return parsedValue;
        }
    }
    if (ncIsNumber(value)) {
        if (value === 1)
            return true;
        if (value === 0)
            return false;
    }
    return null;
};
export const parseJsonValue = (value) => {
    try {
        return ncIsString(value)
            ? JSON.stringify(JSON.parse(value)) // Ensure it's a valid JSON string
            : JSON.stringify(value); // Convert object/array to JSON
    }
    catch (_a) {
        return null;
    }
};
export const parseCurrencyValue = (value, col) => {
    if (ncIsNaN(value)) {
        return null;
    }
    const columnMeta = parseProp(col.meta);
    try {
        return new Intl.NumberFormat(columnMeta.currency_locale || 'en-US', {
            style: 'currency',
            currency: columnMeta.currency_code || 'USD',
        }).format(+value);
    }
    catch (_a) {
        return value;
    }
};
export const parseTimeValue = (value, params) => {
    var _a, _b;
    value = value === null || value === void 0 ? void 0 : value.toString().trim();
    if (!value)
        return null;
    // remove `"`
    // e.g. "2023-05-12T08:03:53.000Z" -> 2023-05-12T08:03:53.000Z
    value = value.replace(/["']/g, '');
    const isMySQL = (_a = params.isMysql) === null || _a === void 0 ? void 0 : _a.call(params, params.col.source_id);
    const isPostgres = (_b = params.isPg) === null || _b === void 0 ? void 0 : _b.call(params, params.col.source_id);
    let d = dayjs(value);
    if (!d.isValid()) {
        // insert a datetime value, copy the value without refreshing
        // e.g. value = 2023-05-12T03:49:25.000Z
        // feed custom parse format
        d = dayjs(value, isMySQL ? 'YYYY-MM-DD HH:mm:ss' : 'YYYY-MM-DD HH:mm:ssZ');
    }
    if (!d.isValid()) {
        // MySQL and Postgres store time in HH:mm:ss format so we need to feed custom parse format
        d = isMySQL || isPostgres ? dayjs(value, 'HH:mm:ss') : dayjs(value);
    }
    if (!d.isValid()) {
        // return empty string for invalid time
        return null;
    }
    return d.format(constructTimeFormat(params.col));
};
export const parseYearValue = (value) => {
    value = value === null || value === void 0 ? void 0 : value.toString().trim();
    if (!value || !dayjs(value).isValid()) {
        return null;
    }
    value = dayjs(value, 'YYYY').format('YYYY');
    return value ? +value : value;
};
export const parseUserValue = (value, withDisplayName = false) => {
    let data = value;
    try {
        if (typeof value === 'string') {
            data = JSON.parse(value);
        }
    }
    catch (_a) { }
    return (ncIsArray(data) ? data : ncIsObject(data) ? [data] : [])
        .map((user) => withDisplayName && user.display_name
        ? `${user.display_name}<${user.email}>`
        : `${user.email}`)
        .join(', ');
};
